Add JSON patching module

particledecay opened the pull request:
##### SUMMARY
Currently, there is no ability to manipulate JSON files. This module adds JSON patching based on [RFC 6902](https://tools.ietf.org/html/rfc6902), with notation conforming to [RFC 6901](https://tools.ietf.org/html/rfc6901) with one notable exception, that I have proposed to the RFC maintainers as well for adoption (RFC 6901 is still in "PROPOSED STANDARD").

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
jsonpatch

##### ADDITIONAL INFORMATION
As a simple example, consider this JSON object in a file named 'test.json':

```json
{
  "foo": "bar"
}
```

The following would add a new member to the JSON:

```yaml
- name: Add or replace 'baz' member
  jsonpatch:
    src: "test.json"
    operations:
      - op: add
        path: "/baz"
        value: "quz"
```

This would result in:

```json
{
  "foo": "bar",
  "baz": "quz"
}
```

In addition to `add`, I have included support for all of the RFC standard operations. For additional information and examples, see [github.com/ParticleDecay/ansible-jsonpatch](https://github.com/ParticleDecay/ansible-jsonpatch). See also [RFC 6901](https://tools.ietf.org/html/rfc6901), [RFC 6902](https://tools.ietf.org/html/rfc6902), and [jsonpatch.com](http://jsonpatch.com/).

particledecay committed:
Add JSON patching module

Fix linting issues in jsonpatch module

Add RETURN block and change some duplicate keys

Updating jsonpatch files to fix errors

feat: multiple changes to json_patch

- rename jsonpatch to json_patch
- use Ansible-native method for backups/writes
- add 'tested' (separate from 'changed') for test operations

fix: remove "_" as variable names

feat: return false for test instead of erroring

feat: support check mode

feat: add option for pretty print JSON

chore(pep8): add blank line

desimaniac commented:
+1 For this

ParticleDecay commented:
Reopening to get ci retested

jamescassell commented on lib/ansible/modules/files/jsonpatch.py:
```diff
@@ -0,0 +1,455 @@
+        """Create a backup copy of the JSON file."""
+        path = self.outfile
+        counter = 0
+        while os.path.exists("%s.bak" % path):
```
Doesn't ansible have a standard way of making backup files?

particledecay replied:
If it does, I missed it in the docs. I'll look at some other modules.

jamescassell commented on lib/ansible/modules/files/jsonpatch.py:
```diff
@@ -0,0 +1,455 @@
+        filename = "%s.bak" % path
+
+        with open(filename, "w") as bak:
+            bak.write(open(self.outfile, "r").read())
```
Does this preserve the modification date of the original file on the backup file?

particledecay replied:
It does not. I'll take a look at how backups are performed in other modules.

jamescassell commented on lib/ansible/modules/files/jsonpatch.py:
```diff
@@ -0,0 +1,455 @@
+            operations=dict(required=True, type='list'),
+            backup=dict(required=False, default=False, type='bool'),
+        ),
+        supports_check_mode=False
```
How hard would it be to add check_mode support?

particledecay replied:
Not sure, since the RFC doesn't do anything like this (and `test` has rather different behavior), but I imagine it wouldn't be terribly hard.

jamescassell reviewed (changes requested):
I'd rename it to json_patch.

Is this module idempotent?  If it's not, can it be?

Would be good to see some integration tests.

ParticleDecay commented:
The module should be idempotent. I just pushed the following:

- renamed from `jsonpatch` to `json_patch`
- switched to Ansible-native method for backups/writes
- add 'tested' (in order to separate from 'changed') as return value for patches that include a test operation

I still need to add check_mode support and maybe some integration tests.

ParticleDecay commented:
This module now supports check mode.

ParticleDecay commented:
Reopening again for stale ci

ParticleDecay commented:
```
ERROR: lib/ansible/modules/files/json_patch.py:0:0: E307 version_added should be '2.9'. Currently '2.8'
```
^ this is entirely due to stale CI (just 3 days ago), so that means this module missed the 2.8 version then?

s-hertel commented:
@ParticleDecay That's correct, the version should be updated to 2.9. It's not clear how to run this module in an idempotent manner - i.e. running the same task always results in changes. Please add integration tests so that can be tested.

particledecay committed:
chore: update version_added for json_patch to 2.9

test: add integration tests for json_patch module

chore: add future imports to json_patch test file

particledecay committed:
fix: address multiple ansibot module errors

- misplaced __future__ import in test_json_patch.py
- missing alias for integration tests
- missing types in various options in DOCUMENTATION string

particledecay committed:
chore: update json_patch for ansible 2.10

particledecay committed:
fix: add test.json file for integration tests

test: add jmespath dependency for json_patch test

fix: add aliases file for jmespath

test: set up pip dependencies within json_patch test

test: add coverage module to json_patch testing

particledecay commented:
@jamescassell @s-hertel ok I'm kinda stuck now. I have integration tests and they pass locally, but the builds aren't passing due to not having "jmespath" installed. I tried adding it as an integration role in `test/integration/targets/jmespath` and making my `json_patch` integration tests dependent on it, but that's not appearing to work either.

I've honestly wasted more time on this than it's worth.

s-hertel commented on test/integration/targets/jmespath/tasks/main.yml:
```diff
@@ -0,0 +1,4 @@
+- name: Install JMESPath
```
Generally for external dependencies tests use a virtual environment rather than modifying the test nodes.
You could add some tasks to set that up like https://github.com/ansible/ansible/blob/924352a051b797b0e8eabec1966b9d985f2ca9b8/test/integration/targets/ecs_cluster/tasks/main.yml, or you could add a runme.sh script like https://github.com/ansible/ansible/blob/0c74ee4352fc5702de25d8bab8a592e0e044f0bc/test/integration/targets/filters/runme.sh that then calls your test playbook.

tpo commented:
Having this module would be **so** appreciated. Oh @particledecay can we make your life less miserable by sending you a few of your favored beers, give you a rush of motivation by having Amazon ship you $YOUR_MOST_DESIRED_BOOK or compensate your pain by sponsoring an hour or two of or your time? If any of that would work then contact me by PM please and I'll see what I can do.

sivel commented:
Hi!

Thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

Unfortunately, we're not sure if we want this feature in the program.

However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at https://docs.ansible.com/ansible/devel/dev_guide/collections_tech_preview.html.

Thank you once again for this and your interest in Ansible!

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

tpo commented:
Hi @sivel,

> Unfortunately, we're not sure if we want this feature in the program.

Then I'd deem it apropriate to publicly document that thinking of those "we". Because:

> Thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

If this phrase is more then a prepared corporate speak template then I think the right way to show respect in Open Source would usually be to send technical feedback and being explicit about rationales etc.

I see that particledecay has "wasted" tons of time on this and doesn't even get a rationale for the "Closed" here? Being explicit about the rationale would also be a courtesy to those that make it to this ticket.

Thanks a lot for your work on ansible!
*t

danbohea commented:
What? Why don't you want this feature in Ansible?

You'll have support for niche things like homebrew cask but not have a sensible way to configure JSON files? Really widely adopted and really popular and OS-agnostic JSON files??

This is ridiculous.

danbohea commented:
Excellent way to discourage anyone from bothering to contribute too - just break out the boilerplate _thanks but no thanks_ reply. I'm really shocked by this result and handling :open_mouth:

particledecay commented:
To be clear: I created this module in the first place because I felt it was an oversight for Ansible to be missing a module for editing JSON files (but could edit things like INI files). I don't think it's particularly reasonable to require Galaxy just so someone can get this functionality. It should be built-in.

I was hoping to also add a YAML module later but to be honest this whole process has been discouraging.

sivel commented:
As an FYI, we are working toward moving all modules into collections, core or otherwise. Community modules we be given back to the community to maintain and publish.

As such we are implementing more strict acceptance of new modules at this time.  As a result from here forward we do not plan to accept new one off community modules.

danbohea commented:
People generously giving their time to develop modules for inclusion in the project could really do with knowing this early on in the process, not towards the end.

tpo commented:
Thanks for the info @sivel 

> As an FYI, we are working toward moving all modules into collections, core or otherwise. Community modules we be given back to the community to maintain and publish.
> 
> As such we are implementing more strict acceptance of new modules at this time. As a result from here forward we do not plan to accept new one off community modules.

Could you please add a link to the rationale (also to remove the nutrients for conspiracies)?

sivel commented:
https://www.ansible.com/blog/thoughts-on-restructuring-the-ansible-project
